x/evm/precompile/oracle.go

```diff
@@ -0,0 +1,130 @@
+package precompile
+
+import (
+	"fmt"
+	"reflect"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	gethabi "github.com/ethereum/go-ethereum/accounts/abi"
+	gethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/vm"
+	gethparams "github.com/ethereum/go-ethereum/params"
+
+	"github.com/NibiruChain/nibiru/v2/app/keepers"
+	"github.com/NibiruChain/nibiru/v2/x/common/asset"
+	"github.com/NibiruChain/nibiru/v2/x/evm/embeds"
+	"github.com/NibiruChain/nibiru/v2/x/evm/statedb"
+	oraclekeeper "github.com/NibiruChain/nibiru/v2/x/oracle/keeper"
+)
+
+var _ vm.PrecompiledContract = (*precompileOracle)(nil)
+
+// Precompile address for "Oracle.sol", the contract that enables queries for exchange rates
+var PrecompileAddr_Oracle = gethcommon.HexToAddress("0x0000000000000000000000000000000000000801")
+
+func (p precompileOracle) Address() gethcommon.Address {
+	return PrecompileAddr_Oracle
+}
+
+func (p precompileOracle) RequiredGas(input []byte) (gasPrice uint64) {
+	// Since [gethparams.TxGas] is the cost per (Ethereum) transaction that does not create
+	// a contract, it's value can be used to derive an appropriate value for the precompile call.
+	return gethparams.TxGas
+}
+
+const (
+	OracleMethod_QueryExchangeRate OracleMethod = "queryExchangeRate"
+)
+
+type OracleMethod string
+
+// Run runs the precompiled contract
+func (p precompileOracle) Run(
+	evm *vm.EVM, contract *vm.Contract, readonly bool,
+) (bz []byte, err error) {
+	// This is a `defer` pattern to add behavior that runs in the case that the error is
+	// non-nil, creating a concise way to add extra information.
+	defer func() {
+		if err != nil {
+			precompileType := reflect.TypeOf(p).Name()
+			err = fmt.Errorf("precompile error: failed to run %s: %w", precompileType, err)
+		}
+	}()
+
+	// 1 | Get context from StateDB
+	stateDB, ok := evm.StateDB.(*statedb.StateDB)
+	if !ok {
+		err = fmt.Errorf("failed to load the sdk.Context from the EVM StateDB")
+		return
+	}
+	ctx := stateDB.GetContext()
+
+	method, args, err := DecomposeInput(embeds.SmartContract_Oracle.ABI, contract.Input)
+	if err != nil {
+		return nil, err
+	}
+
+	switch OracleMethod(method.Name) {
+	case OracleMethod_QueryExchangeRate:
+		// TODO: UD-DEBUG: Test that calling non-method on the right address does
+		// nothing.
+		bz, err = p.queryExchangeRate(ctx, method, args, readonly)
+	default:
+		// TODO: UD-DEBUG: test invalid method called
```

Comment:
```suggestion
		bz, err = p.queryExchangeRate(ctx, method, args, readonly)
	default:
```

I later discovered that both scenarios described by those TODO comments are impossible because of the Input.Unpack logic inside of `DecomposeInput`. The `switch` block is relatively safe, so long as we have a case for each method defined in the precompile's ABI.